crates/pathfinder/src/merkle_tree.rs

```diff
@@ -0,0 +1,1089 @@
+//! trees with the same leaves must have the same nodes). The path of an edge node can therefore
+//! be many bits long.
+//!
+//! [Binary nodes](Node::Binary) is a branch node with two children, left and right. This represents
+//! only a single bit on the path to a leaf.
+//!
+//! A tree storing a single key-value would consist of two nodes. The root node would be an edge node
+//! with a path equal to the key. This edge node is connected to a leaf node storing the value.
+//!
+//! #### Implementation details
+//!
+//! We've defined an additional node type, an [Unresolved node](Node::Unresolved). This is used to
+//! represent a node who's hash is known, but has not yet been retrieved from storage (and we therefore
+//! have no further details about it).
+//!
+//! Our implementation is a mix of nodes from persistent storage and any mutations are kept in-memory. It is
+//! done this way to allow many mutations to a tree before committing only the final result to storage. This
+//! may be confusing since we just said trees are immutable -- but since we are only changing the in-memory
+//! tree, the immutable tree still exists in storage. One can therefore think of the in-memory tree as containing
+//! the state changes between tree `N` and `N + 1`.
+//!
+//! The in-memory tree is built using a graph of [Rc<RefCell<Node>>] which is a bit painful.
+
+mod node;
+mod storage;
+
+use anyhow::Context;
+use bitvec::prelude::BitVec;
+use rusqlite::Transaction;
+use std::{cell::RefCell, rc::Rc};
+
+use crate::merkle_tree::{
+    node::{BinaryNode, Direction, EdgeNode, Node},
+    storage::{PersistedBinaryNode, PersistedEdgeNode, PersistedNode, RcNodeStorage},
+};
+
+use pedersen::StarkHash;
+
+pub const ZERO_HASH: StarkHash = StarkHash::zero();
+
+/// A Starknet binary Merkle-Patricia tree with a specific root entry-point and storage.
+///
+/// This is used to update, mutate and access global Starknet state as well as individual contract states.
+///
+/// For more information on how this functions internally, see [here](super::merkle_tree).
+#[derive(Debug, Clone)]
+pub struct MerkleTree<'a> {
+    storage: RcNodeStorage<'a>,
+    root: Rc<RefCell<Node>>,
+}
+
+impl<'a> MerkleTree<'a> {
+    /// Removes one instance of the tree and its root from persistent storage.
+    ///
+    /// This implies decrementing the root's reference count. The root will
+    /// only get deleted if the reference count reaches zero. This will in turn
+    /// delete all internal nodes and leaves which no longer have a root to connect to.
+    ///
+    /// This allows for multiple instances of the same tree state to be committed,
+    /// without deleting all of them in a single call.
+    pub fn delete(self) -> anyhow::Result<()> {
+        match self.root.borrow().hash() {
+            Some(hash) if hash != ZERO_HASH => self
+                .storage
+                .decrement_ref_count(hash)
+                .context("Failed to delete tree root"),
```

Comment:
Not crucial right now, but in such cases context could provide the hash val that caused the error for convenience.

Reply:
The caller also has access to the root he tried to delete.